segments/weather.sh

```diff
@@ -3,51 +3,64 @@
-# The data provider to use. Currently only "yahoo" is supported.
+# The data provider to use. Currently only "yrno" is supported.
 export TMUX_POWERLINE_SEG_WEATHER_DATA_PROVIDER="${TMUX_POWERLINE_SEG_WEATHER_DATA_PROVIDER_DEFAULT}"
 # What unit to use. Can be any of {c,f,k}.
 export TMUX_POWERLINE_SEG_WEATHER_UNIT="${TMUX_POWERLINE_SEG_WEATHER_UNIT_DEFAULT}"
 # How often to update the weather in seconds.
 export TMUX_POWERLINE_SEG_WEATHER_UPDATE_PERIOD="${TMUX_POWERLINE_SEG_WEATHER_UPDATE_PERIOD_DEFAULT}"
-# Name of GNU grep binary if in PATH, or path to it.
-export TMUX_POWERLINE_SEG_WEATHER_GREP="${TMUX_POWERLINE_SEG_WEATHER_GREP_DEFAULT}"
-# Location of the JSON parser, jq
-export TMUX_POWERLINE_SEG_WEATHER_JSON="${TMUX_POWERLINE_SEG_WEATHER_JSON_DEFAULT}"
+# How often to update the weather location in seconds (this is only used when latitude and longitude settings are set to "auto")
+export TMUX_POWERLINE_SEG_WEATHER_LOCATION_UPDATE_PERIOD="${TMUX_POWERLINE_SEG_WEATHER_LOCATION_UPDATE_PERIOD_DEFAULT}"
 # Your location
 # Latitude and Longtitude for use with yr.no
-# Set both to "auto" to detect automatically based on your IP address
-TMUX_POWERLINE_SEG_WEATHER_LAT=""
-TMUX_POWERLINE_SEG_WEATHER_LON=""
+# Set both to "auto" to detect automatically based on your IP address, or set them manually
+export TMUX_POWERLINE_SEG_WEATHER_LAT="${TMUX_POWERLINE_SEG_WEATHER_LAT_DEFAULT}"
+export TMUX_POWERLINE_SEG_WEATHER_LON="${TMUX_POWERLINE_SEG_WEATHER_LON_DEFAULT}"
 EORC
 	echo "$rccontents"
 }
 
 run_segment() {
-	__process_settings
-	local tmp_file="${TMUX_POWERLINE_DIR_TEMPORARY}/temp_weather_file.txt"
-	local weather
-	case "$TMUX_POWERLINE_SEG_WEATHER_DATA_PROVIDER" in
-	"yrno") weather=$(__yrno) ;;
-	*)
-		echo "Unknown weather provider [$TMUX_POWERLINE_SEG_WEATHER_DATA_PROVIDER]"
-		return 1
-		;;
-	esac
-	if [ -n "$weather" ]; then
-		echo "$weather"
+	local cache_weather_file="${TMUX_POWERLINE_DIR_TEMPORARY}/weather_cache_data.txt"
+	local cache_location_file="${TMUX_POWERLINE_DIR_TEMPORARY}/weather_cache_location.txt"
+	local weather=""
+
+	# Check if the weather data is still a valid cache hit
+	if [ -f "$cache_weather_file" ]; then
+		last_update=$(__read_file_last_update "$cache_weather_file")
+		time_now=$(date +%s)
+		up_to_date=$(echo "(${time_now}-${last_update}) < ${TMUX_POWERLINE_SEG_WEATHER_UPDATE_PERIOD}" | bc)
+		if [ "$up_to_date" -eq 1 ]; then
+			weather=$(__read_file_content "$cache_weather_file")
+		fi
 	fi
+	
+	# Fetch from provider if empty
+	# If a new provider is implemented, please set the $weather variable!
+	if [ -z "$weather" ]; then
+		__process_settings
+		case "$TMUX_POWERLINE_SEG_WEATHER_DATA_PROVIDER" in
+		"yrno")
+			weather=$(__yrno)
+			;;
+		*)
+			# Just read the stale cache, default to empty/blank
+			# Better not overwriting the previous healthy content
+			weather=$(__read_file_content "$cache_weather_file")
+			;;
+		esac
+		echo "$weather" > "$cache_weather_file"
+	fi
+
+	echo "$weather"
```

Comment:
Wondrful 👏 🥳 🎆 , now this is the only echo for all code paths. This is a fantastic simplication of this old messy segment. Well don @onlyuuuuu !